### PR TITLE
KAFKA-4930: Added connector name validator …

### PR DESCRIPTION
**Pull request by soenkeliebau**
…to check for empty connector name and illegal characters in connector name. This also fixes  KAFKA-4938 by removing the check for slashes in connector name from ConnectorsResource.

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5802/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5816/
Test PASSed (JDK 7 and Scala 2.11).

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 1]
```diff
@@ -909,6 +909,39 @@ public void ensureValid(String name, Object o) {
         }
     }
 
+    public static class ValidName implements Validator {
```
Comment (gwenshap) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: I don't think these are general rules (i.e. do they apply to client IDs? Streams application IDs?). So perhaps move this to the ConnectorConfig class?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 2]
```diff
@@ -86,18 +86,40 @@ public ConnectorsResource(Herder herder) {
...
         Map<String, String> configs = createRequest.config();
-        if (!configs.containsKey(ConnectorConfig.NAME_CONFIG))
-            configs.put(ConnectorConfig.NAME_CONFIG, name);
+        CreateConnectorRequest finalCreateRequest = createRequest;
+
+        String nameFromBody = createRequest.name();
+        String nameFromConfig = createRequest.config().get(ConnectorConfig.NAME_CONFIG);
+        String name = "";
+
+        if (nameFromBody == null && nameFromConfig == null) {
```
Comment (gwenshap) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java]: I'm a bit confused - if we have the validator, why do we need to add all the checks here?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResourceTest.java [hunk 3]
```diff
@@ -193,11 +193,32 @@ public void testCreateConnectorExists() throws Throwable {
         PowerMock.verifyAll();
     }
 
+    // Requests to create a connector that do not contain a name in either the body or the config object should be
+    // rejected outright
     @Test(expected = BadRequestException.class)
-    public void testCreateConnectorWithASlashInItsName() throws Throwable {
-        String badConnectorName = CONNECTOR_NAME + "/" + "test";
+    public void testCreateConnectorWithoutName() throws Throwable {
```
Comment (gwenshap) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResourceTest.java]: This is an empty name, not null, right? Maybe clarify, since the difference is meaningful.

**Review by gwenshap — CHANGES_REQUESTED** [review]
Hey, few small comments. 

Can you also add a test for "bad characters"?

If the existing test is for empty name, maybe add a test to validate that null is fine.

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java (hunk 1 above)
Comment (soenkeliebau) [reply clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Makes sense, I'll move to ConnectorConfig class.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java (hunk 2 above)
Comment (soenkeliebau) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java]: Those tests might be due to a misunderstanding on my part, back when I looked at this, it was possible to send a request without a top level "name" object, so I figured that it was legal to specify the name within the config object of the request. The current version does not accept requests without the top level "name" element, so I'll remove these tests.

I put them here, because there could potentially have been two connector names and we can only pass one along to the config object, so they were not possible to do in the validator.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResourceTest.java (hunk 3 above)
Comment (soenkeliebau) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResourceTest.java]: Will clarify

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java [hunk 4]
```diff
@@ -909,6 +909,39 @@ public void ensureValid(String name, Object o) {
...
+            if (s == null) {
+                // This can happen during creation of the config object due to no default value being defined for the
+                // name configuration - a missing name parameter is caught when checking for mandatory parameters,
+                // thus we can ok a null value here
+                return;
+            } else if (s.isEmpty()) {
+                throw new ConfigException(name, value, "Name may not be empty");
+            }
+
+            // Check name string for illegal characters
```
Comment (viktorsomogyi) [inline clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Hi Soenke, I'd also make a comment if you don't mind.
Have you considered using precompiled regexes to match against the connector names?

Comment (soenkeliebau) [reply clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Hi Viktor, I considered it, yes, but decided on this way because it saves us from building a regex using string operations, which to my mind always bears the potential of escaping issues. 
The alternative would be to directly have the regexp in the code, which tbh I find considerably less readable than the way it currently is.
I figured that performance is less of an issue for this, since adding a connector will not be called that often, but to be on the safe side I just did some quick'n dirty tests on performance and a precompiled regex was actually a bit slower than the indexOf way. But I will grant that there may be some tuning potential in the regex code there.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java (hunk 2 above)
Comment (soenkeliebau) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java]: When working on this I rediscovered what made me write these conditions, which was these two lines from [ConnectorsResource.java](https://github.com/apache/kafka/blob/e31c0c9bdbad432bc21b583bd3c084f05323f642/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java#L95):

`if (!configs.containsKey(ConnectorConfig.NAME_CONFIG))
            configs.put(ConnectorConfig.NAME_CONFIG, name);`

These make me think that the name parameter at the top level of the request and the name parameter in the config sub-element are the same thing.
With this code, it is possible to create something like the following:

```
 {
	"name": "testconn2",
	"config": {
		"name": "badname",
		"connector.class": "org.apache.kafka.connect.tools.MockSourceConnector",
		"tasks.max": "1",
		"topics": "test-topic"
	},
	"tasks": []
}
```
Which may or may not be what is wanted..if a connector requires a name config and that needs to be different from the connector name this may be useful, but it is also prone to misunderstandings I think.

I am tempted to leave a check in place that disallows passing in two different name elements.
Any thoughts?

File: clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java (hunk 4 above)
Comment (viktorsomogyi) [reply clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java]: Ok, I think I'm fine with your arguments, thank you for explaining :).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5905/
Test FAILed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5890/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by soenkeliebau** [comment]
retest this please

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/5900/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/5915/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by soenkeliebau** [comment]
*bump* 
@gwenshap Did you have a chance to take a look at the changes?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java [hunk 5]
```diff
@@ -252,4 +254,37 @@ public boolean visible(String name, Map<String, Object> parsedConfig) {
...
+                // This can happen during creation of the config object due to no default value being defined for the
+                // name configuration - a missing name parameter is caught when checking for mandatory parameters,
+                // thus we can ok a null value here
+                return;
+            } else if (s.isEmpty()) {
+                throw new ConfigException(name, value, "Name may not be empty");
+            }
+
+            // Check name string for illegal characters
+            ArrayList<Character> foundIllegalCharacters = new ArrayList<>();
```
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java]: Should this maybe be a `Set` instead, so that invalid characters used more than once are only reported once?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java [hunk 6]
```diff
@@ -252,4 +254,37 @@ public boolean visible(String name, Map<String, Object> parsedConfig) {
...
+        public void ensureValid(String name, Object value) {
+            String s = (String) value;
+
+            if (s == null) {
+                // This can happen during creation of the config object due to no default value being defined for the
+                // name configuration - a missing name parameter is caught when checking for mandatory parameters,
+                // thus we can ok a null value here
+                return;
+            } else if (s.isEmpty()) {
+                throw new ConfigException(name, value, "Name may not be empty");
```
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java]: What about "Connector name may not be empty" or "Connector name must be supplied"?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 7]
```diff
@@ -86,10 +86,11 @@ public ConnectorsResource(Herder herder) {
     @POST
     @Path("/")
     public Response createConnector(final @QueryParam("forward") Boolean forward,
-                                    final CreateConnectorRequest createRequest) throws Throwable {
+                                     final CreateConnectorRequest createRequest) throws Throwable {
```
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java]: Minor nit: incorrect indentation.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 8]
```diff
@@ -86,10 +86,11 @@ public ConnectorsResource(Herder herder) {
...
         String name = createRequest.name();
-        if (name.contains("/")) {
-            throw new BadRequestException("connector name should not contain '/'");
+        if (name == null) {
+            // the request did not contain a connector name, reject it as a bad request
+            throw new BadRequestException("connector name is mandatory in create connector requests");
```
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java]: What about "connector name must be supplied"?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ConnectorConfigTest.java [hunk 9]
```diff
@@ -69,7 +69,7 @@ public void close() {
         @Override
         public ConfigDef config() {
             return new ConfigDef()
-                    .define("magic.number", ConfigDef.Type.INT, ConfigDef.NO_DEFAULT_VALUE, ConfigDef.Range.atLeast(42), ConfigDef.Importance.HIGH, "");
+                .define("magic.number", ConfigDef.Type.INT, ConfigDef.NO_DEFAULT_VALUE, ConfigDef.Range.atLeast(42), ConfigDef.Importance.HIGH, "");
```
Comment (rhauch) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ConnectorConfigTest.java]: Minor nit: it'd be good to not include unnecessary changes like this.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResourceTest.java (hunk 3 above)
Comment (rhauch) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResourceTest.java]: Wouldn't it also be useful to keep the `testCreateConnectorWithInvalidName()` while also adding `testCreateConnectorWithoutName()`?

**Review by rhauch — COMMENTED** [review]
Thanks for the suggested changes. I have a few comments detailed below.

**Comment by soenkeliebau** [comment]
Thanks for your comments @rhauch. I have been out of commission for a few weeks, but will review and address these tonight.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java (hunk 6 above)
Comment (soenkeliebau) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java]: Agreed, it should be specified that the connector name is referenced here. Will change to second suggestion.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java (hunk 5 above)
Comment (soenkeliebau) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectorConfig.java]: Absolutely, will change!

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java (hunk 7 above)
Comment (soenkeliebau) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java]: Will change.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java (hunk 8 above)
Comment (soenkeliebau) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java]: Will change to be consistent with the other message above.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ConnectorConfigTest.java (hunk 9 above)
Comment (soenkeliebau) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/ConnectorConfigTest.java]: That was one of those cases of Intellj meaning well and me not catching it, will revert.

**Comment by soenkeliebau** [comment]
@rhauch Thanks for your comments, I have addressed most of them. Regarding the last comment on removing the test case (for some reason github won't let me reply to that directly, so I'll add my thoughts here) - in my commit I move the test for illegal characters from ConnectorsResource to ConnectorConfig,  which the tests in ConnectorsResourceTest mock, so the test case did not make much sense there anymore. The test should now be covered by [validNames](https://github.com/apache/kafka/pull/2755/files#diff-41aea8c2d0bb713644dab70ebad11ef0R173) in ConnectorConfigTest. @

**Comment by asfgit** [comment]
SUCCESS 
 8070 tests run, 5 skipped, 0 failed.
--none--

**Comment by soenkeliebau** [comment]
retest this please

**Comment by asfgit** [comment]
SUCCESS 
 8074 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8074 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8074 tests run, 5 skipped, 0 failed.
--none--

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java [hunk 10]
```diff
@@ -87,10 +87,10 @@ public ConnectorsResource(Herder herder) {
     @Path("/")
     public Response createConnector(final @QueryParam("forward") Boolean forward,
                                     final CreateConnectorRequest createRequest) throws Throwable {
-        String name = createRequest.name();
-        if (name.contains("/")) {
```
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:91]: So it looks like we now remove the check for `/`. Just want to check, the assumption is that `/` is escaped now, right? Is the implication that KAFKA-4372 was incorrect and that the particular DELETE endpoint just needed to use some additional encoding to ensure it was possible to delete the connector (presumably with every other endpoint that has a connector name in the URL)?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResourceTest.java [hunk 11]
```diff
@@ -66,12 +66,26 @@
     private static final String LEADER_URL = "http://leader:8083/";
     private static final String CONNECTOR_NAME = "test";
     private static final String CONNECTOR2_NAME = "test2";
+    private static final String CONNECTOR_NAME_ALL_WHITESPACES = "     ";
```
Comment (ewencp) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResourceTest.java]: to get a bit better coverage on type of whitespace instead of just spaces, might be worth throwing in a few other whitespace characters into these examples.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResourceTest.java [hunk 12]
```diff
@@ -196,20 +210,62 @@ public void testCreateConnectorExists() throws Throwable {
         PowerMock.verifyAll();
     }
 
-    @Test(expected = BadRequestException.class)
-    public void testCreateConnectorWithASlashInItsName() throws Throwable {
-        String badConnectorName = CONNECTOR_NAME + "/" + "test";
+    @Test
+    public void testCreateConnectorNameTrimWhitespaces() throws Throwable {
+        // Clone CONNECTOR_CONFIG_WITHOUT_NAME Map, as createConnector changes it (puts the name in it) and this
```
Comment (ewencp) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResourceTest.java:237]: this might be a good reason to make it a `static` method to get an instance of it instead of making it a `static` field. it could be easy to misuse if left as a field.

this is a good catch. i'd suggest making a copy in `createConnector` but since it's the entry point for a REST call it shouldn't ever be necessary.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResourceTest.java [hunk 13]
```diff
@@ -196,20 +210,62 @@ public void testCreateConnectorExists() throws Throwable {
...
+        // will affect later tests
+        Map<String, String> inputConfig = new HashMap<>();
+        inputConfig.putAll(CONNECTOR_CONFIG_WITHOUT_NAME);
+        final CreateConnectorRequest bodyIn = new CreateConnectorRequest(CONNECTOR_NAME_PADDING_WHITESPACES, inputConfig);
+        final CreateConnectorRequest bodyOut = new CreateConnectorRequest(CONNECTOR_NAME, CONNECTOR_CONFIG);
```
Comment (ewencp) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResourceTest.java]: doesn't seem this is necessary? couldn't we just pass the parameters here directly to `putConnectorConfig` and the `ConnectorInfo` constructor? same with `bodyOut` in tests below.

**Review by ewencp — COMMENTED** [review]
@soenkeliebau a few comments in the code and just one comment in tests. i think based on KIP discussion this will be updated a little bit more, but it seems very close.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java (hunk 10 above)
Comment (soenkeliebau) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResource.java:91]: Yes. Deleting connectors with / in their name works fine when percent encoding the slash in the url.
I believe this check is no longer needed as with the proposed implementation we can handle slashes in connector names (if they are percent encoded when being used as part of a url, in a json string they can be used as is). Additionally this check creates a NullPointerException when the name parameter is missing from a request (see KAFKA-4938).

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResourceTest.java (hunk 11 above)
Comment (soenkeliebau) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResourceTest.java]: Will do

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResourceTest.java (hunk 12 above)
Comment (soenkeliebau) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResourceTest.java:237]: You are right, I'll move the cloning code to a static method.

**Comment by ewencp** [comment]
@soenkeliebau This looked very, very close to ready to ship and we're just about at the KIP deadline. Any chance we can wrap up the last few comments so we can get this committed?

**Comment by soenkeliebau** [comment]
Hi @ewencp - I actually already had the window open to start work on this in the background, just had to finish up a work thing first. If I don't stumble across anything unforeseen I should hopefully commit a new version in 2 - 3 hours.

**Comment by ewencp** [comment]
@soenkeliebau Awesome! ping me when it's ready for final review an commit

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResourceTest.java (hunk 13 above)
Comment (soenkeliebau) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/rest/resources/ConnectorsResourceTest.java]: Not sure I fully understand this comment tbh. Do you mean we just need testPutConnectorConfig and skip the extra tests that actually call createConnector?

**Comment by soenkeliebau** [comment]
@ewencp I just pushed a new version, hopefully all checks should pass, locally the connect tests all passed.
Due to the discussions around KIP-212 this has changed a bit since your last review, I have added a new validator that rejects control characters and also added tests for that validator to the existing tests.

**Comment by ewencp** [comment]
@soenkeliebau Cool, LGTM. I updated with a minor checkstyle fix, will just wait for tests.

The `ConnectorsResourceTest`s are fine since they are correct (we expect those invalid values to be forwarded to the herder because we don't do validation in the Resource). We might want to add tests of the invalid characters in names to the `ConnectorConfigTest` similar to the empty name, though I'm not too concerned since we have the validator on the config field and the validator has tests.

**Comment by soenkeliebau** [comment]
retest this please

**Comment by soenkeliebau** [comment]
The H24 build server seems to be out of disk space. I don't have access to the infra mailing lists to let them know, but I'll assume this will be fixed soon. So will retry in a few hours.

**Comment by ewencp** [comment]
@soenkeliebau no worries, the builder that has failures is unrelated and we're good on the other builds, so this LGTM, merging to trunk for 1.1.0
